### PR TITLE
Feature: save and load measurements

### LINKED ISSUE #52: Create mapper from FHIR Observations to Health Connect Records
### Problem

Currently, a mapper has been created in the [healthconnectonfhir](https://github.com/StanfordSpezi/SpeziKt/tree/main/modules/healthconnectonfhir) module that converts Health Connect Records into FHIR Observations. However, no mapper exists to convert FHIR Observations back into Health Connect Records.

### Solution

As discussed in our team meeting today, we will create a mapper that takes FHIR Observations and maps them to Health Connect Records by mapping the LOINC code of the Observation to the corresponding Record subclass.

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct and Contributing Guidelines

### PR DESCRIPTION
**Pull request by Basler182**
# *save and load measurements*

## :gear: Release Notes 
image

- save measurements
- load latest measurements
- Fixes #52 

## :white_check_mark: Testing

as soon as it is confirmed that the functionality fits, further tests will be added

## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt
Comment (PSchmiedmayer) [inline app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt:72]: If we pull collations, we should look into snapshot listeners to ensure that we always get the latest set of values in case something changes: https://firebase.google.com/docs/firestore/query-data/listen#kotlin+ktx

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt
Comment (PSchmiedmayer) [inline app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt:84]: Might make sense to rather rely on a snapshot listener + the overall list of collections and get the first element from there rather than multiple Firebase queries.

**Review by PSchmiedmayer — COMMENTED** [review]
Nice job with the functionality @Basler182!

From a pure UX perspective: As discussed in our meeting today, it would be great to replace "Measured on: 30 Juni" with a concrete date and time, e.g. "June 30, 2:42 PM". In addition, the horizontal and vertical distance between the boxes should be the same.

 Nice job with the overall structure. I only had one high-level comment to use snapshot listeners for regular updates in the app; I will leave a detailed code review to @eldcn 🚀

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt
Comment (Basler182) [reply app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt:72]: Thanks for the tip At this point it seems like it might just be overhead because the new measurements only come from the app right now. However, I will be happy to try this out with the messages feature, as the new data comes from outside the app and it will make it much easier to stay up to date at all times and when we notice something.

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/models/BluetoothUiState.kt
Comment (eldcn) [inline app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/models/BluetoothUiState.kt:20]: 💪 

Very nice. Relates to #39 and it's the correct way to handle loading states

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/component/VitalDisplay.kt
Comment (eldcn) [inline app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/component/VitalDisplay.kt:55]: This is again complex logic to be handled in compose. The ui states should already serve the data in such a way to be consumed directly from composables. It is the task of the mappers to construct the texts. See the existing example in `BluetoothUiStateMapper`. Instead of OperationStatus enum, I would use a statusText in the ui state instead, and `measuredOnText: String?`

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/component/VitalDisplay.kt
Comment (eldcn) [inline app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/component/VitalDisplay.kt:66]: As mentioned above, this text should be constructed in the mapper. Keep also in mind string resources so that we can handle localization as well

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/mapper/MeasurementToRecordMapper.kt
Comment (eldcn) [inline app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/mapper/MeasurementToRecordMapper.kt:14]: As proposed above, this functionalities can be merged in the main mapper of the view model

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/models/VitalDisplayUiState.kt
Comment (eldcn) [inline app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/models/VitalDisplayUiState.kt:5]: This object shouldn't be needed, I would use the three properties (VitalDisplayData) directly in the main ui state of the view model instead for each type

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt
Comment (eldcn) [inline app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt:104]: this is not needed/used

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/screen/MeasurementDialog.kt
Comment (eldcn) [inline app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/screen/MeasurementDialog.kt:1]: Consider defining strings in this file and throughout this PR in general in string resources

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/screen/MeasurementDialog.kt
Comment (eldcn) [inline app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/screen/MeasurementDialog.kt:1]: You can suppress the factory instead

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/screen/MeasurementDialog.kt
Comment (eldcn) [inline app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/screen/MeasurementDialog.kt:64]: These texts should be served from the ui state and constructed in the mapper

File: modules/healthconnectonfhir/src/main/java/edu/stanford/healthconnectonfhir/Loinc.kt
Comment (eldcn) [inline modules/healthconnectonfhir/src/main/java/edu/stanford/healthconnectonfhir/Loinc.kt:3]: I would be more in favor of an enum instead `enum class Loinc(code: String)`, and move `COMPONENT` in the `ObservationToRecordMapperImpl` in companion object

File: modules/healthconnectonfhir/src/main/java/edu/stanford/healthconnectonfhir/ObservationToRecordMapperImpl.kt
Comment (eldcn) [inline modules/healthconnectonfhir/src/main/java/edu/stanford/healthconnectonfhir/ObservationToRecordMapperImpl.kt:13]: We can we mark this and `RecordToObservationMapperImpl` to internal

File: modules/healthconnectonfhir/src/main/java/edu/stanford/healthconnectonfhir/ObservationToRecordMapperImpl.kt
Comment (eldcn) [inline modules/healthconnectonfhir/src/main/java/edu/stanford/healthconnectonfhir/ObservationToRecordMapperImpl.kt:21]: I would introduce a generic type here:

```kotlin
   override fun <T : Record> map(observation: Observation): T {
        @Suppress("UNCHECKED_CAST")
        return when(observation.code.codingFirstRep.code) {
            Loinc.BloodPressure.CODE -> mapToBloodPressureRecord(observation)
            Loinc.Weight.CODE -> mapToWeightRecord(observation)
            Loinc.HeartRate.CODE -> mapToHeartRateRecord(observation)
            else -> error("")
        } as T
    }
```

and reflect the same generic type in the observation repository as well

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt
Comment (eldcn) [inline app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt:109]: Not needed the overhead imo, you can use `observationToRecordMapper.map(observation)` directly

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt
Comment (eldcn) [inline app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt:126]: Can be marked as private and instead of code: String, I would use the proposed enum for `Loinc` instead, type-safety ftw

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt
Comment (eldcn) [inline app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt:113]: The component is called observation repository, methods as well getXXXObservation, but these methods return records. I would avoid mapping to records at this component and do that only in view model via the main mapper of the view model

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt
Comment (eldcn) [inline app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt:58]: Minor: we can move this before forEach so that we only create one instance

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt
Comment (eldcn) [inline app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt:56]: Question: Should we would wrapp the content of each item here in a runCatching so that we don't miss saving the whole batch of observations in case of a single invalid one?

**Review by eldcn — COMMENTED** [review]
The issue that we are facing now it is related to this one: https://github.com/hapifhir/hapi-fhir/issues/3268

The version lib version that we are using contains a huge file (over 80k lines) that can't be instrumented and transformed from Jacoco. I would disable androidTest coverage in app module for now and to unblock this PR and we address it separately

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/BluetoothViewModel.kt
Comment (Basler182) [reply app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/BluetoothViewModel.kt:43]: Yeah, good idea! :)

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/screen/MeasurementDialog.kt
Comment (Basler182) [reply app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/screen/MeasurementDialog.kt:1]: good idea!

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/screen/MeasurementDialog.kt
Comment (Basler182) [reply app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/screen/MeasurementDialog.kt:1]: good point! 👍

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt
Comment (Basler182) [reply app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt:109]: well seen :)

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt
Comment (Basler182) [reply app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/repository/ObservationRepository.kt:56]: Theoretically, the method could also be used to store any number of independent observations. In our case, however, it is the case that related data is always saved, where it may be more intuitive not to save the others (in our case so far a maximum of 1 more) if one fails.

File: modules/healthconnectonfhir/src/main/java/edu/stanford/healthconnectonfhir/Loinc.kt
Comment (Basler182) [reply modules/healthconnectonfhir/src/main/java/edu/stanford/healthconnectonfhir/Loinc.kt:3]: but don't we then lose the connection of which Loinc code can have which components?

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/MeasurementModule.kt
Comment (Basler182) [reply app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/MeasurementModule.kt:17]: you are right, makes sense! 👍

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/BluetoothViewModel.kt
Comment (Basler182) [reply app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/BluetoothViewModel.kt:35]: here i'm not quite sure, because not all mappers have anything to do with the ui state. in the ui state we show the "measurement" and some mappers we only use to map the measurement into a record or observation so that we can save it, but don't use it for visualisation. or would you still include it in the ui state mapper, even if it's not directly related to the ui state, @eldcn ?

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/BluetoothViewModel.kt
Comment (Basler182) [reply app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/BluetoothViewModel.kt]: after talking to @PSchmiedmayer , i have slightly changed the logic. we no longer just query the data, but have a listener that notifies us of changes. i have made the optimisations with the mapper. i like it, then the viemodel will be a little bit slimmer again :)

File: modules/healthconnectonfhir/src/main/java/edu/stanford/healthconnectonfhir/ObservationToRecordMapperImpl.kt
Comment (Basler182) [reply modules/healthconnectonfhir/src/main/java/edu/stanford/healthconnectonfhir/ObservationToRecordMapperImpl.kt:21]: Good idea. I like this approach. 💯

File: modules/healthconnectonfhir/src/main/java/edu/stanford/healthconnectonfhir/ObservationToRecordMapperImpl.kt
Comment (Basler182) [reply modules/healthconnectonfhir/src/main/java/edu/stanford/healthconnectonfhir/ObservationToRecordMapperImpl.kt:13]: I think this is not possible cause in `HealthConnectOnFhirModule`  it would say that public function exposes its internal parameter type.

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/component/VitalDisplay.kt
Comment (Basler182) [reply app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/component/VitalDisplay.kt:61]: but without the enum it feels even more complex. then we have to create logic when e.g. the CircularProgressBar is displayed if e.g. Status.isEmpty and text.isEmpty, etc?

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/mapper/MeasurementToRecordMapper.kt
Comment (Basler182) [reply app/src/main/kotlin/edu/stanford/bdh/engagehf/bluetooth/data/mapper/MeasurementToRecordMapper.kt]: logic moved into repository

**Comment by eldcn** [comment]
@PSchmiedmayer Unfortunately, the fixed that I added still didn't work as there was a crash at runtime for missing classpath. We decided to disable coverage for android tests for now, and will think for an alternative solution in a follow up. 

Also, there will be a follow up to address some smaller issues to not block this PR. 

Since the code coverage decreased, the code-cov job fails, and we might need your support to merge directly. This iteration looks very good so far 👍

**Review by eldcn — APPROVED** [review]

**Comment by Basler182** [comment]
@PSchmiedmayer and we can make the changes due to the firestore changes in a small follow up pr, since not much changes, it should be done quickly

**Comment by PSchmiedmayer** [comment]
@Basler182 @eldcn Sounds good, thank you for all the work here! 

Please note any remaining issues in GitHub issues and let me know when I should merge the PR.

Also fine to address the changes in paths in a separate, smaller PR as long is both systems don't stay out of sync for too long.

**Comment by Basler182** [comment]
> @Basler182 @eldcn Sounds good, thank you for all the work here!
> 
> Please note any remaining issues in GitHub issues and let me know when I should merge the PR.
> 
> Also fine to address the changes in paths in a separate, smaller PR as long is both systems don't stay out of sync for too long.

we are ready to go. you are welcome to merge it, @PSchmiedmayer  🚀

**Comment by PSchmiedmayer** [comment]
Sounds great; thank you for the work @Basler182 @eldcn!

**Pull request merged**
